src/nfa/truffle_simd.hpp

```diff
@@ -53,6 +59,166 @@ const SuperVector<S> blockSingleMask(SuperVector<S> shuf_mask_lo_highclear, Supe
 #include "ppc64el/truffle.hpp"
 #endif
 
+#ifdef HAVE_SVE
+
+const u8 *truffleExecReal(m128 shuf_mask_lo_highclear, m128 shuf_mask_lo_highset,
+                      const u8 *buf, const u8 *buf_end);
+
+const u8 *rtruffleExecReal(m128 shuf_mask_lo_highclear, m128 shuf_mask_lo_highset,
+                       const u8 *buf, const u8 *buf_end);
+
+static really_inline
+const u8 *scanBlock(svuint8_t shuf_mask_lo_highclear, svuint8_t shuf_mask_lo_highset, svuint8_t chars, const u8 *buf, bool forward) {
+
+    const size_t vector_size_int_8 = svcntb();
+
+    const svuint8_t result_mask = blockSingleMask(shuf_mask_lo_highclear, shuf_mask_lo_highset, chars);
+    uint64_t index;
+    if (forward) {
+        index = first_non_zero(vector_size_int_8, result_mask);
+    } else {
+        index = last_non_zero(vector_size_int_8, result_mask);
+    }
+
+    if(index < vector_size_int_8) {
+        return buf+index;
+    } else {
+        return NULL;
+    }
+}
+
+const u8 *truffleExecReal(m128 shuf_mask_lo_highclear, m128 shuf_mask_lo_highset, const u8 *buf, const u8 *buf_end) {
+    const int vect_size_int8 = svcntb();
+    // Activate only 16 lanes to read the m128 buffers
+    const svbool_t lane_pred_16 = svwhilelt_b8(0, 16);
+    assert(buf && buf_end);
+    assert(buf < buf_end);
+    DEBUG_PRINTF("truffle %p len %zu\n", buf, buf_end - buf);
+    DEBUG_PRINTF("b %s\n", buf);
+
+    svuint8_t wide_shuf_mask_lo_highclear = svld1(lane_pred_16, (uint8_t*) &shuf_mask_lo_highclear);
+    svuint8_t wide_shuf_mask_lo_highset = svld1(lane_pred_16, (uint8_t*) &shuf_mask_lo_highset);
+
+    const u8 *work_buffer = buf;
+    const u8 *ret_val;
+
+    DEBUG_PRINTF("start %p end %p \n", work_buffer, buf_end);
+    assert(work_buffer < buf_end);
+
+    __builtin_prefetch(work_buffer +   64);
```

Comment:
I have started an prefetch optimization and I've found that in some cases I got up to 50% faster performance in the microbenchmarks: https://github.com/VectorCamp/vectorscan/commit/ab36731b9dd5c3ea26f4870654b258f159a85a66
Could you test these changes for SVE as well and confirm they are also valid in this PR?

Reply:
I did not observe any difference on my machine with the legacy and the new prefetch, so I set it the same way you did.
baseline (neon): Truffle: 5 matches, 16000 * 62500 iterations, total elapsed time = 0.308 s, average bandwidth = 4935.293 MB/s
patched with default prefetch: Truffle: 5 matches, 16000 * 62500 iterations, total elapsed time = 0.259 s, average bandwidth = 5860.817 MB/s
patched with new prefetch: Truffle: 5 matches, 16000 * 62500 iterations, total elapsed time = 0.259 s, average bandwidth = 5858.518 MB/s